### DISCUSSION TITLE
Problem blocking AD ID permission in kid-targeted app

### DISCUSSION THREAD

**Discussion by mgravano** [airsdk/Adobe-Runtime-Support#1588] [Q&A]
* AIR SDK version: 33.1.1.743
* Affected environment: Android

Hello, Starting April 1, 2022, Google Play will require apps directed at children not to use Ad ID. For that, it is necessary to include the following element in the applications manifest file to prevent the permission from being combined with your application:

`<uses-permission android:name="com.google.android.gms.permission.AD_ID" tools:node="remove"/>`

The problem is that when I include it it doesn't allow me to publish my app. It tells me "the application descriptor file cannot be parsed".
Does anyone know of a way to include this element without causing this error?

Thanks!

**Comment by marchbold** [comment]
The `tools` namespace allows you to manipulate the manifest from other libraries that may have added the permission. the "remove" part is just ensuring that this permission is removed from the manifest. 

As AIR doesn't add permissions automatically (unless you are using apm) then this is not needed, just don't add the permission to your manifest additions. 

If you are using apm you can create a custom android manifest addition (https://github.com/airsdk/apm/wiki/Usage-Generate#android):

```
apm generate config android
```

Then edit `config/android/AndroidManifest.xml` as follows:

```xml
<?xml version="1.0" encoding="utf-8"?>
<manifest xmlns:android="http://schemas.android.com/apk/res/android" xmlns:tools="http://schemas.android.com/tools">

    <uses-permission android:name="com.google.android.gms.permission.AD_ID" tools:node="remove"/>

    <!-- other custom manifest additions -->

</manifest>
```

Then when you generate the manifest additions it will remove this permission if it was added from any of the extensions you have in use.

```
apm generate app-descriptor src/MyApp-app.xml
```

**Comment by mgravano** [comment]
Thank you for your reply Michael!

Are you sure that if I don't include the AD_ID permission the app won't send the AD ID anyway? Sorry for asking, but on this Google page (https://support.google.com/googleplay/android-developer/answer/6048248#zippy=%2Cinfracciones-de-la-pol%C3%ADtica-del-id-de-publicidad) says the following:

**"Some SDKs, such as the Google Mobile Ads SDK (play-services-ads), may already declare this permission in their library's manifest file. If your app uses these SDKs as dependencies, the AD_ID permission in the SDK library's manifest file will be merged into the main app manifest file by default, even if you don't explicitly declare the permission in your application main manifest file.

If your app uses SDKs, such as the Google Mobile Ads SDK (which declares the AD_ID permission in its library manifest files), you must include the following element in your manifest file to prevent the permission from being combined with your application.

`<uses-permission android:name="com.google.android.gms.permission.AD_ID" tools:node="remove"/>`

This permission will be enforced on all devices starting in early 2022."**

On the other hand, what is apm?

**Reply by ajwfrost** [reply]
So, we can adjust things a little so that you're able to add that `tools:node="remove"` request - you'd need to also define the namespace within the `manifest` section though e.g.:
```
	<android>
		<manifestAdditions><![CDATA[
			<manifest xmlns:tools="http://schemas.android.com/tools">
				<uses-permission android:name="com.google.android.gms.permission.AD_ID" tools:node="remove"/>
			</manifest>
		]]></manifestAdditions>
	</android>
```

apm = AIR Package Manager (think like NPM for JavaScript?) - https://github.com/airsdk/apm/wiki
It helps configure dependencies and app descriptors etc..

thanks

**Reply by mgravano** [reply]
Thank you Andrew!
I tried what you told me and I get the following error:

![error](https://user-images.githubusercontent.com/5115296/150682390-a97359fb-de14-423c-9ecc-38f5177303b6.jpg)

I tried these two ways:

`<manifest xmlns:tools="http://schemas.android.com/tools">`

and:

`<manifest xmlns:tools="http://schemas.android.com/tools" android:installLocation="auto">`

I include all my code so you can see if there are any errors:

![code](https://user-images.githubusercontent.com/5115296/150682914-26abc5f3-7920-49c4-b123-6c2202b1cb67.jpg)

Thank you!

**Reply by ajwfrost** [reply]
Yes, sorry @mgravano, I had meant that we'll need to also tweak the code here. I just checked and for some reason there's some filtering going on when writing the final manifest file to disk, which is overriding all URI namespaces to change them to `android`. So the `tools:node` bit is being replaced by `android:node` and this isn't recognised hence the gradle error.

If you generate the Android Studio project, you can manually change it in the manifest file, but we'll get a fix out soon to correct this so that it's handled better...

thanks

**Reply by marchbold** [reply]
@mgravano Yes positive if you don't add this to your manifest additions AIR doesn't add it! You don't need to wait for this. 

The remove node is used in normal android development as they automatically merge in the manifest from the libraries in use. Without the "tools remove" addition the AD ID permission may be added from a dependency. As this doesn't happen in AIR (currently) this is not needed.

apm is the AIR Package Manager: 
- https://airsdk.dev/docs/basics/install-apm
- https://github.com/airsdk/apm/wiki

It can perform manifest merging from installed packages which may ad the AD ID but if you don't use it then this won't be happening.  :)

**Reply by mgravano** [reply]
Understood. Thank you very much Michael!

**Comment by mgravano** [comment]
So I'll have to wait for you guys to fix it. Do you have any idea when the change will be ready?
Thanks

**Reply by ajwfrost** [reply]
There's one other fix we're trying to sort out for Android which will hopefully go in early this week, so aiming for a release maybe Monday 31st all going well... (or, I can upload an interim version of the ADT tool which might help if you need this more quickly?)

**Reply by mgravano** [reply]
Hello, just to know, the update with the solution to this problem has not been published yet, right?

**Reply by ajwfrost** [reply]
We just posted the updated release today but only as a beta - there are other changes that are a bit more wide-ranging so you have to go looking for it if you want to use this version:
https://airsdk.harman.com/download/33.1.1.758
thanks

**Reply by mgravano** [reply]
Thank you Andrew! It seems to have worked, I had no problem publishing the app.

**Comment by mgravano** [comment]
Monday the 31st is fine. I'll wait for it, thanks!
